drivers/spi/spi_silabs_eusart.c

```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+
+	/* invoke callback at the end of transfer list. */
+	current_dma_config->complete_callback_en = 0x00U;
+
+	current_dma_config->source_data_size = 0x01U; /* 1 byte. */
+	current_dma_config->dest_data_size = 0x01U;   /* 1 byte. */
+
+	current_dma_config->source_burst_length = 0x01U;
+	current_dma_config->dest_burst_length = 0x01U;
+
+	current_dma_config->block_count = block_count;
+
+	current_dma_config->head_block = (struct dma_block_config *)blk_cfg;
+
+	current_dma_config->user_data = (void *)dev;
+
+	if (!is_tx) {
+		/* Only enable the end transaction callback. Only SPI controller mode is supported.
+		 */
+		current_dma_config->dma_callback = &spi_silabs_dma_rx_callback;
+	}
+}
+
+static int spi_silabs_prepare_dma_transfer(const struct device *dev,
+					   size_t total_padded_transaction_size, bool is_tx)
+{
+	struct dma_block_config *current_blk_cfg = NULL;
+	const struct spi_buf *current_buffer = NULL;
+	struct spi_silabs_eusart_data *instance_data = dev->data;
+	const struct spi_silabs_eusart_config *instance_conf = dev->config;
+
+	struct dma_config dma_cfg;
+
+	size_t current_total_transfer_size = 0U;
+	size_t padding_size = 0U;
+	size_t buffer_count = 0U;
+	size_t dma_block_config_idx = 0U;
+	int ret = 0U;
+
+	if (is_tx) {
+		current_buffer = &instance_data->ctx.current_tx[0];
+		buffer_count = instance_data->ctx.tx_count;
+		current_blk_cfg = &(instance_data->dma_channel_tx.dma_descriptors[0]);
+	} else {
+		current_buffer = &instance_data->ctx.current_rx[0];
+		buffer_count = instance_data->ctx.rx_count;
+		current_blk_cfg = &(instance_data->dma_channel_rx.dma_descriptors[0]);
+	}
+
+	/* For each buffer in the instance context buffer set. */
+	for (size_t i = 0U; i < buffer_count; i++) {
+		size_t buffer_remaining_count = current_buffer->len;
+
+		current_total_transfer_size += current_buffer->len;
+
+		if (buffer_remaining_count == 0U) {
+			/* Skip zero length buffers */
+			continue;
+		}
+
+		/* Create the required number of dma transfer block for this buffer. */
+		while (dma_block_config_idx < SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			uint32_t index_in_buffer = current_buffer->len - buffer_remaining_count;
+
+			buffer_remaining_count = spi_silabs_build_dma_transfer_block(
+				instance_conf, &(current_blk_cfg[dma_block_config_idx]),
+				&(((uint8_t *)current_buffer->buf)[index_in_buffer]),
+				buffer_remaining_count, is_tx);
+
+			dma_block_config_idx++;
+
+			if (dma_block_config_idx > SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+				LOG_ERR("DMA channel transfer blocks exausted. Can't prepare the "
+					"transaction.");
+				return -ENOMEM;
+			}
+
+			if (buffer_remaining_count == 0U) {
+				current_buffer++;
+				break;
+			}
+		}
+	}
+
+	padding_size = total_padded_transaction_size - current_total_transfer_size;
+
+	while (padding_size) {
+
+		padding_size = spi_silabs_build_dma_transfer_block(
+			instance_conf, &current_blk_cfg[dma_block_config_idx], NULL, padding_size,
```

Comment:
`dma_block_config_idx` could already be `== SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL` here, so this access is an overflow. The check on line 463 needs to happen at the beginning of the loop, before the variables are accessed.

Reply:
I see this was modified, but now `dma_block_config_idx` is never incremented in the loop, so it will fail if more than 1 padding block is needed.